### ISSUE TITLE
Type mismatch of 'memref.store' when lowering std to llvm

### ISSUE THREAD

**Issue by tjuwhy** [llvm/llvm-project#57310]
code :
```mlir
//example.mlir
module  {
  func @main() {
    %cst_4 = constant 1.000000e+00 : f64
    %2 = memref.alloc() : memref<2x3xf64>
    %c0 = constant 0 : index
    %c0_5 = constant 0 : index
    memref.store %cst_4, %2[%c0, %c0_5] : memref<2x3xf64>    
    memref.dealloc %2 : memref<2x3xf64>
    return
  }
}
```
I tried to run `mlir-opt -convert-std-to-llvm example.mlir` and got the following error:
```console
example.mlir:7:5: error: 'memref.store' op operand #2 must be index, but got 'i64'
    memref.store %cst_4, %2[%c0, %c0_5] : memref<2x3xf64>
    ^
example.mlir:7:5: note: see current operation: "memref.store"(%0, %1, %2, %3) : (f64, memref<2x3xf64>, i64, i64) -> ()
```

It looks like the lowering pass covert the type of`%c0` and `%c0_5` from index to i64, while the operand `memref.store` is not passed to adapt this change

**Comment by rajashekarkasturi** [comment]
Just trying to restructure your code, prefix "arith" before using constant : `arith.constant`
Try this!
```
module  {
  func @main() {
    %cst_4 = arith.constant 1.000000e+00 : f64
    %2 = memref.alloc() : memref<2x3xf64>
    %c0 = arith.constant 0 : index
    %c0_5 = arith.constant 0 : index
    memref.store %cst_4, %2[%c0, %c0_5] : memref<2x3xf64>
    memref.dealloc %2 : memref<2x3xf64>
    return
  }
}
```

```
> mlir-opt example.mlir --convert-std-to-llvm --convert-memref-to-llvm
```

**Comment by tjuwhy** [comment]
Thanks for your suggestion.

Unfortunately, I received errors as follows:
```
affine-github.mlir:3:14: error: custom op 'arith.constant' is unknown
    %cst_4 = arith.constant 1.000000e+00 : f64
```

It is strange because I did pull the latest source of llvm-project and there did exist the Arithmetic Dialect.

I suspect this is due to my mlir being built based on the old version llvm. As `mlir-opt --version` says:
```
mlir-opt --version
LLVM (http://llvm.org/):
  LLVM version 13.0.0
  Optimized build with assertions.
  Default target: x86_64-unknown-linux-gnu
  Host CPU: broadwell
```

**Comment by rajashekarkasturi** [comment]
Yeah, try to checkout latest if possible.
FYI, I'm using
```
> mlir-opt --version
LLVM (http://llvm.org/):
  LLVM version 14.0.6
  Optimized build with assertions.
  Default target: x86_64-unknown-linux-gnu
  Host CPU: broadwell
```

**Comment by tjuwhy** [comment]
Thanks, it works! After I updated LLVM to the latest release. Now I can lower the example to the LLVM dialect through `mlir-opt example.mlir --convert-std-to-llvm --convert-memref-to-llvm > llvm.mlir`. 

But still, the `llvm.mlir` contains unrealized operations, for example:
``` 
%61 = "builtin.unrealized_conversion_cast"(%60) : (i64) -> index
```
so that I cannot further lower the LLVM dialect to LLVM IR.

I wonder if there is another way to lower down the high level dialects to the LLVM dialect? If there isn't, how does IREE does it?

**Comment by rajashekarkasturi** [comment]
Yeah, use the flag --reconcile-unrealized-casts
```
> mlir-opt example.mlir --convert-std-to-llvm --convert-memref-to-llvm --reconcile-unrealized-casts > llvm.mlir
```

**Comment by tjuwhy** [comment]
Thanks, I think my problem is solved

**Comment by EugeneZelenko** [comment]
@tjuwhy: Could you please elaborate?

**Comment by tjuwhy** [comment]
Yes, sorry for the delay, I spend some time confirming my issue. 

When I tried `mlir-opt --reconcile-unrealized-casts arith.llvmcc.mlir`, it raised an error.   
```
arith.llvmc.mlir:21:11: error: failed to legalize operation 'builtin.unrealized_conversion_cast' that was explicitly marked illegal
    %17 = builtin.unrealized_conversion_cast %16 : i64 to index
          ^
```
As discussed above, the `unrealized_conversion_cast is supposed` to be eliminated through the flag. But it failed to.


The 'arith.llvmcc.mlir' can be found [here](https://drive.google.com/file/d/10NFjrqARFO6C95VPft98Q0m7_fNq7ve2/view?usp=sharing), again sorry for didn't provide the reduced file since I'm not familiar with mlir-reduce. More info I can provide is, the value `%17` is used by
```
%354 = memref.atomic_rmw minf %297, %194[%17, %17] : (f32, memref<3x3xf32>) -> f32
```
If I remove all the usage  of `%17` (by removing all the operations followed by `%17 = ...`), the error is eliminated. Please give me some suggestions.

**Comment by tjuwhy** [comment]
> Yes, sorry for the delay, I spend some time confirming my issue.
> 
> When I tried `mlir-opt --reconcile-unrealized-casts arith.llvmcc.mlir`, it raised an error.
> 
> ```
> arith.llvmc.mlir:21:11: error: failed to legalize operation 'builtin.unrealized_conversion_cast' that was explicitly marked illegal
>     %17 = builtin.unrealized_conversion_cast %16 : i64 to index
>           ^
> ```
> 
> As discussed above, the `unrealized_conversion_cast is supposed` to be eliminated through the flag. But it failed to.
> 
> The 'arith.llvmcc.mlir' can be found [here](https://drive.google.com/file/d/10NFjrqARFO6C95VPft98Q0m7_fNq7ve2/view?usp=sharing), again sorry for didn't provide the reduced file since I'm not familiar with mlir-reduce. More info I can provide is, the value `%17` is used by
> 
> ```
> %354 = memref.atomic_rmw minf %297, %194[%17, %17] : (f32, memref<3x3xf32>) -> f32
> ```
> 
> If I remove all the usage of `%17` (by removing all the operations followed by `%17 = ...`), the error is eliminated. Please give me some suggestions.

@EugeneZelenko Sorry to bother you, could you please give me some advices？

**Comment by EugeneZelenko** [comment]
@tjuwhy: Sorry, I'm not code maintainer. I just so that you closed issue but without details why this was done.
